src/day8.scala

```
@@ -0,0 +1,136 @@
+  case Two extends Digit(A, C, D, E, G)
+  case Three extends Digit(A, C, D, F, G)
+  case Four extends Digit(B, C, D, F)
+  case Five extends Digit(A, B, D, F, G)
+  case Six extends Digit(A, B, D, E, F, G)
+  case Seven extends Digit(A, C, F)
+  case Eight extends Digit(A, B, C, D, E, F, G)
+  case Nine extends Digit(A, B, C, D, F, G)
+
+object Digit:
+
+  private val uniqueLookup: Map[Int, Digit] =
+    values.groupBy(_.segments.size).collect { case k -> Array(d) => k -> d }
+
+  def lookupUnique(segments: Segments): Option[Digit] =
+    uniqueLookup.get(segments.size)
+
+end Digit
+
+def part1(input: String): Int =
+
+  def getDisplay(line: String): String = line.split("\\|")(1).trim
+
+  val parseUniqueDigit: String => Option[Digit] =
+    Segment.parseSeq andThen Digit.lookupUnique
+
+  val uniqueDigits: Iterator[Digit] =
+    for
+      display <- input.linesIterator.map(getDisplay)
+      segments <- display.split(" ")
+      uniqueDigit <- parseUniqueDigit(segments)
+    yield
+      uniqueDigit
+
+  uniqueDigits.length
+end part1
+
+def part2(input: String): Int =
+
+  def parseSegmentsSeq(segments: String): Seq[Segments] =
+    segments.trim.split(" ").toSeq.map(Segment.parseSeq)
+
+  def splitParts(line: String): (Seq[Segments], Seq[Segments]) =
+    val Array(cypher, plaintext) = line.split("\\|").map(parseSegmentsSeq)
+    (cypher, plaintext)
+
+  def digitsToInt(digits: Seq[Digit]): Int =
+    digits.foldLeft(0)((acc, d) => acc * 10 + d.ordinal)
+
+  val problems = input.linesIterator.map(splitParts)
+
+  val solutions = problems.map((cypher, plaintext) =>
+    plaintext.map(substitutions(cypher))
+  )
+
+  solutions.map(digitsToInt).sum
+
+end part2
+
+def substitutions(cypher: Seq[Segments]): Segments => Digit =
+
+  def lookup(section: Seq[Segments], withSegments: Set[Segment]): (Segments, Seq[Segments]) =
+    val found = section.find(s => (s.toSet & withSegments) == withSegments).get
+    val remaining = section.filterNot(_ == found)
```

Consider using `partition` instead:
```suggestion
    val (Seq(uniqueMatch), remaining) = section.partition(s => withSegments.subsetOf(s))
    (uniqueMatch, remaining)
```
This also has the advantage of failing early if there are two or more inputs matching the given segments.

---

I like this!